src/main/kotlin/com/github/avrokotlin/avro4k/Avro.kt

```diff
@@ -19,262 +19,314 @@ import kotlinx.serialization.SerializationStrategy
+
+
+    /**
+     * Writes an instance of <T> using a [Schema] derived from the type.
+     * This method will use the [AvroEncodeFormat.Data] format without a codec.
+     * The written object will be returned as a [ByteArray].
+     */
+    override fun <T> encodeToByteArray(serializer: SerializationStrategy<T>, value: T): ByteArray {
+        val baos = ByteArrayOutputStream()
+        openOutputStream(serializer) {
+            encodeFormat = AvroEncodeFormat.Data()
+        }.to(baos).write(value).close()
+        return baos.toByteArray()
+    }
+
+    /**
+     * Creates an [AvroOutputStreamBuilder] that will write instances of <T>.
+     * Supply a function to this method to configure the builder, eg
+     *
+     * <pre>
+     * val output = openOutputStream<T>(serializer) {
+     *    encodeFormat = AvroEncodeFormat.Data()
+     *    schema = mySchema
+     * }
+     * </pre>
+     *
+     * If the schema is not supplied in the configuration function then it will
+     * be derived from the type using [Avro.schema].
+     */
+    fun <T> openOutputStream(
+            serializer: SerializationStrategy<T>,
+            f: AvroOutputStreamBuilder<T>.() -> Unit = {}
+    ): AvroOutputStreamBuilder<T> {
+        val builder = AvroOutputStreamBuilder(serializer, this) { schema -> { encode(serializer, schema, it) } }
+        builder.f()
+        return builder
+    }
+
+    /**
+     * Converts an instance of <T> to an Avro [Record] using a [Schema] derived from the type.
+     */
+    @Deprecated("Has been replaced by #encode that is able to not-only encode records", replaceWith = ReplaceWith("this.encode(serializer, obj)"))
+    fun <T> toRecord(
+            serializer: SerializationStrategy<T>,
+            obj: T,
+    ): GenericRecord {
+        return toRecord(serializer, schema(serializer), obj)
+    }
+
+    /**
+     * Converts an instance of <T> to an Avro [Record] using the given [Schema].
+     */
+    @Deprecated("Has been replaced by #encode that is able to not-only encode records", replaceWith = ReplaceWith("this.encode(serializer, schema, obj)"))
+    fun <T> toRecord(serializer: SerializationStrategy<T>,
+                     schema: Schema,
+                     obj: T): GenericRecord {
+        val encodedValue = encode(serializer, schema, obj)
+        if (encodedValue !is GenericRecord)
+            throw AvroRuntimeException("Expected a GenericRecord, found $encodedValue")
+        return encodedValue
+    }
+
+    fun <T> encode(serializer: SerializationStrategy<T>, value: T): Any? =
+            encode(serializer, schema(serializer), value)
+
+    fun <T> encode(serializer: SerializationStrategy<T>,
+                   schema: Schema,
+                   value: T): Any? {
+        val encoder = GenericAvroEncoder(schema, serializersModule, configuration)
+        encoder.resolveSchema(serializer.descriptor, value == null)
+        serializer.serialize(encoder, value)
+        return encoder.encodedValue
+    }
```

Comment:
What has changed in this file is only the handling of all value types in root, nothing else